corehq/apps/sms/models.py

```diff
@@ -2295,4 +2296,111 @@ def has_migration_completed(cls, name):
             return False
 
 
+class Keyword(SyncSQLToCouchMixin, models.Model):
+    """
+    A Keyword allows a project to define actions to be taken when a contact
+    in the project sends an inbound SMS starting with a certain word.
+    """
+
+    class Meta:
+        index_together = (
+            ('domain', 'keyword')
+        )
+
+    couch_id = models.CharField(max_length=126, null=True, db_index=True)
+    domain = models.CharField(max_length=126, db_index=True)
+
+    # The word which is used to invoke this Keyword's KeywordActions
+    keyword = models.CharField(max_length=126)
+    description = models.TextField(null=True)
+
+    # When specified, this is the delimiter that is used in the structured SMS format.
+    # If None, the delimiter is any consecutive white space.
+    # This is ignored unless this Keyword is describing a structured SMS
+    # (i.e., it has a KeywordAction with action equal to ACTION_STRUCTURED_SMS)
+    delimiter = models.CharField(max_length=126, null=True)
+
+    # If a SQLXFormsSession is open for a contact when they invoke this Keyword,
+    # override_open_sessions tells what to do with it. If True, the SQLXFormsSession
+    # will be closed and this Keyword will be invoked. If False, this Keyword will be
+    # skipped and the form session handler will count the text as the next
+    # answer in the open survey.
+    override_open_sessions = models.NullBooleanField()
+
+    # List of doc types representing the only types of contacts who should be
+    # able to invoke this keyword. Empty list means anyone can invoke.
+    # Example: ['CommCareUser', 'CommCareCase']
+    initiator_doc_type_filter = jsonfield.JSONField(default=list)
+
+
+class KeywordAction(models.Model):
+    """
+    When a Keyword is invoked, its KeywordActions are processed. A KeywordAction
+    defines the action to take (which could be sending an SMS, or starting an
+    SMS survey, for example) and the recipient of that action.
+    """
+
+    # Send an SMS
+    ACTION_SMS = "sms"
+
+    # Start an SMS Survey
+    ACTION_SMS_SURVEY = "survey"
+
+    # Process the text as structured SMS. The expected format of the structured
+    # SMS is described using the fields on this object.
+    ACTION_STRUCTURED_SMS = "structured_sms"
+
+    # The recipient of this action is the contact who invoked the keyword.
+    RECIPIENT_SENDER = "SENDER"
+
+    # The recipient of this action is the owner of the case contact who invoked
+    # the keyword.
+    RECIPIENT_OWNER = "OWNER"
+
+    # The recipient of this action is a user group (Group) with id given by
+    # recipient_id.
+    RECIPIENT_USER_GROUP = "USER_GROUP"
+
+    # The Keyword that this KeywordAction belongs to
+    keyword = models.ForeignKey('Keyword', on_delete=models.CASCADE)
+
+    # One of the ACTION_* constants representing the action to take
+    action = models.CharField(max_length=126)
```

Comment:
I've stopped using `choices` because django doesn't actually enforce that one of those values are selected before you save the model, and it doesn't add a constraint on the db side either in the migration. 

I think `choices` is only used on the UI side if you're using a `ModelForm` to create/edit the object. It does provide self-documentation of the choices though, but the other problem with it is every time you add a new choice, django prompts you that there are changes which require a migration, so we'd have to follow up each change to the choices list with a migration.

The 126 is because for a lot of these I'm not really sure what the max length should be, so I just go with the max possible length while having postgres still consider all strings entered to be "short strings". (see [here](https://www.postgresql.org/docs/9.3/static/datatype-character.html))